### PR TITLE
feat(accountsdb): implement account cache

### PR DESCRIPTION
**Pull request by Sobeston**
aims to close https://github.com/Syndica/sig/issues/281

### CODE REVIEW

File: src/accountsdb/db.zig
Comment (Sobeston) [inline src/accountsdb/db.zig:371]: Should fix and re-enable these tests before merging. There's a memory leak somewhere

File: src/accountsdb/db.zig
Comment (0xNineteen) [inline src/accountsdb/db.zig]:
```suggestion
        if (slot_lru.put(pubkey, new_cached_account) != null) return error.AlreadyExistsInCache;
```

File: src/accountsdb/db.zig
Comment (0xNineteen) [inline src/accountsdb/db.zig:184]:
```suggestion
    fn shiftLRUForward(self: *AccountsLRU, old_slot: Slot, new_slot: Slot) !void {
```

File: src/accountsdb/db.zig
Comment (0xNineteen) [inline src/accountsdb/db.zig:206]: is this always true?

**Review by 0xNineteen — CHANGES_REQUESTED** [review]
awesome work - it took a bit to understand the architecture but its very well designed and once it clicked, very cool too - most comments are related to organizing the code to how we have it elsewhere in the codebase and naming

[for anyone else doing a PR review - this diagram may help:

Screenshot 2024-10-10 at 9 57 17 PM

File: src/accountsdb/db.zig
Comment (Sobeston) [reply src/accountsdb/db.zig:206]: Good point, it might not be

File: src/accountsdb/cache.zig
Comment (Sobeston) [inline src/accountsdb/cache.zig:311]: There's a memory leak hiding here somewhere, will re-enable this test when found

Comment (Sobeston) [reply src/accountsdb/cache.zig]: Fixed - slot_lru.put was overwriting the old value in cases where the key already existed, causing the value to be dangling and not freed. Fixed by introducing putNoClobber.

**Review by 0xNineteen — DISMISSED** [review]
just pushed a commit to fix some naming update changes - lgtm :shipit: 🔥

should be good to merge once davids/inks review is done

File: src/accountsdb/cache.zig
Comment (InKryption) [inline src/accountsdb/cache.zig:86]: What's the point of returning a `CachedAccount` here, if not as a a pointer, and without `copyRef`? This seems like a potential source of race conditions. Please correct me if I'm off the mark here.

File: src/accountsdb/cache.zig
Comment (InKryption) [inline src/accountsdb/cache.zig:56]: Could you generally comment on the multi-threaded interaction with the AccountsCache itself, and each LRU entry, and/or any future plans regarding this?

Within AccountsDB the whole account cache is under a read-write lock, so I'm struggling to see when having each individual LRU entry be thread-safe in and of itself would be necessary.

File: src/accountsdb/cache.zig
Comment (InKryption) [inline src/accountsdb/cache.zig:188]: To this and all other instances of this in the other tests: I think it would be worth falling in line with #304, which looks like it's going to be merged.

**Review by InKryption — CHANGES_REQUESTED** [review]
A couple of minor nits, and a couple of questions about the design related to multi-threaded interactions I'd like to understand and potentially document.

I overall quite like the design, I think it's a clear improvement.

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig:56]: Down the line I'd like to get rid of the mutex wrapping the cache and make it internally locking.

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig]: https://github.com/Syndica/sig/pull/297/commits/ae36aa27390e13793a692a6c0f2c1356439b4e27#diff-7f62328c8d0fd51e206b91ff2b72aef34de0fae176ba46106a0c768d29f306f8L17-L19

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig]: https://github.com/Syndica/sig/pull/297/commits/ae36aa27390e13793a692a6c0f2c1356439b4e27#diff-7f62328c8d0fd51e206b91ff2b72aef34de0fae176ba46106a0c768d29f306f8L86-R104

**Review by InKryption — APPROVED** [review]

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:41]: There are multiple races here. I'm not going to list them, but here's how it should look:
```suggestion
            const current_count = self.ref_count.fetchSub(1, .release);
            if (current_count == 1) {
                _ = self.ref_count.load(.acquire);
                self.deinit(allocator);
                allocator.destroy(self);
            }
```

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:50]: This should be `monotonic`.

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:57]: These should be `monotonic`.

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:102]: Maybe split this into a 
```zig
const account = slot_lru.get(pubkey) orelse return null;
return account.copyRef();
```
I feel like that would look cleaner. Same for the one below.

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:174]: This should be `monotonic`.

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:195]: Not a blocker, just a suggestion:
``` suggestion
                lowest_slot = @min(entry.key_ptr.*, lowest_slot);
```
Maybe it's clearer? Curious to hear your opinion.

**Review by Rexicon226 — CHANGES_REQUESTED** [review]

File: src/accountsdb/cache.zig
Comment (dnut) [inline src/accountsdb/cache.zig:19]: Every time you want to mark an account dirty, you need to get it from the Lru. This simultaneously marks the account as the most recently used item, and also as a stale item at the same time. Something feels off about that.

**Review by dnut — CHANGES_REQUESTED** [review]
There are some potential races with the reference counter. I suggest using `sig.sync.ReferenceCounter` if you want a robust way to count references. That counter is a bit more paranoid than necessary for this scenario, since you already have deinits guarded with an exclusive write lock, but I think it's better to share a common rc rather than reinvent the wheel in various different places.

Also I'm still not sure about about indexing starting from the Slot. Seems like it brings unnecessary overhead.

My comments go into more detail.

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig:117]: I think I worded it too strongly, it's fine if the caller calls it on an existing key, it will just not modify the state and return an error.

File: src/accountsdb/cache.zig
Comment (Rexicon226) [reply src/accountsdb/cache.zig:50]: Potential solution: assert the `ref_count` isn't 0

**Comment by Rexicon226** [comment]
> There are some potential races with the reference counter. I suggest using sig.sync.ReferenceCounter if you want a robust way to count references. That counter is a bit more paranoid than necessary for this scenario, since you already have deinits guarded with an exclusive write lock, but I think it's better to share a common rc rather than reinvent the wheel in various different places.

I do want to note that `ReferenceCounter` is not correctly implemented as of now. I'll open a PR to fix it later. Just use it for now, it won't affect too much.

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig]: Removed per-slot behaviour in https://github.com/Syndica/sig/pull/297/commits/f039a412be0a120833bb2e60244d6b958c526707

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig]: changed in https://github.com/Syndica/sig/pull/297/commits/ee1d3869d31b9a1d9ffd9b4bec8a186a5d3a4fd6#diff-7f62328c8d0fd51e206b91ff2b72aef34de0fae176ba46106a0c768d29f306f8L11-R20

File: src/accountsdb/db.zig
Comment (0xNineteen) [inline src/accountsdb/db.zig:1881]: read lock?

**Review by 0xNineteen — DISMISSED** [review]
lgtm

File: src/accountsdb/db.zig
Comment (Sobeston) [reply src/accountsdb/db.zig:1881]: The .get requires a mutable pointer, as it can modify the lru state (previously the lru state was behind a pointer).

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:49]: This panic doesn't make sense to have, `ref_count` will also be destroyed with the last release.

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:41]: This is a hack to fix something that we ourselves wrote. `LruCacheCustom` should be patched so as not to need this.

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:104]: This allocation seems superfluous to me. The LRU already allocates new keys when they are added. Can we rewrite something somewhere to not need this second allocation on every `put` instead of on the first one?

**Review by Rexicon226 — CHANGES_REQUESTED** [review]
It looks great. From a maintainability point of view, I have a couple of nits that I would like to address before merging.

File: src/accountsdb/cache.zig
Comment (dnut) [inline src/accountsdb/cache.zig:66]: This is redundant with the RwMux. I get that you want internal locking for this structure eventually, but specifying `locking` here isn't good enough for that, which is why you need the RwMux. It's not clear to me how a `locking` Lru could be beneficial here, so I would just make this `non_locking` for now. If you want to start moving towards internal locking, you can support it fully by integrating a RwLock in AccountsCache, and get rid of the RwMux wrapper.

**Review by dnut — CHANGES_REQUESTED** [review]
Looks good, I just have a few more thoughts

File: src/accountsdb/cache.zig
Comment (dnut) [reply src/accountsdb/cache.zig:41]: Might be a complicated patch to retain compatibility with the situation where you're storing values instead of pointers in the cache. If your other suggestion is implemented, I don't think we need this function any more

> This allocation seems superfluous to me. The LRU already allocates new keys when they are added. Can we rewrite something somewhere to not need this second allocation on every put instead of on the first one?

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig:113]: The LRU calls that whenever it deinits a node so that's already handled

File: src/accountsdb/cache.zig
Comment (dnut) [reply src/accountsdb/cache.zig:49]: I agree that the only case where this could theoretically be false is when the pointer itself is already invalidated and unsafe to use. If the function is called at that point, it means you already have a bug somewhere else, from mismanaging the pointer, and the behavior of calling this function is undefined.

So you could just have `_ = self.ref_count.acquire()` here. Essentially you use it as a basic reference counter, and don't bother to check for cases that are actually impossible to accurately detect.

That being said, maybe there is an advantage to having a panic or unreachable here. Nothing is guaranteed, since any code that could trigger this would have undefined behavior, but at least you get a probabilistic tilt towards failing fast instead of failing silently.

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig:104]: The reasoning is that the LRU doesn't necessarily "own" the account, we don't necessarily want the entry to be deallocated when it's evicted; references to it could still be held. With that being said we could remove one allocation per put by allocating a CachedAccount and its .account field in the same block, but that feels like a bit of a hack.

Comment (dnut) [reply src/accountsdb/cache.zig:104]: I think the core issue here is that the reference counter must be allocated in some memory that is not owned by the lru. It's fine to copy the CachedAccount struct into the Lru but it needs to respect the same reference counter that's being used by other copies of the CachedAccount.

So you need an allocation outside the Lru regardless, it's just a question of whether to allocate the CachedAccount as a whole, or to only allocate its reference counter.

File: src/accountsdb/db.zig
Comment (Sobeston) [reply src/accountsdb/db.zig]: Addressed in https://github.com/Syndica/sig/pull/297/commits/de424828b6dbad6b49ebf2b09db33a614f237941

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig]: I went ahead and changed this to an unreachable, I think conveys intent a bit better

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig]: I went ahead and replaced it with an unreachable, I think that shows intent a bit better

File: src/accountsdb/cache.zig
Comment (Sobeston) [reply src/accountsdb/cache.zig]: Simplified with https://github.com/Syndica/sig/pull/297/commits/4ad2cb1e2a5f853c1041c03d46dc013536737e76

File: src/common/lru.zig
Comment (Sobeston) [reply src/common/lru.zig:45]: Yeah that sounds right to me. Might as well fix the other spot too, let me get that.

**Review by dnut — APPROVED** [review]
@Rexicon226 have all your concerns been addressed?

File: src/accountsdb/cache.zig
Comment (Rexicon226) [inline src/accountsdb/cache.zig:105]: This function isn't actually used anywhere, plus users can just call `self.lru.contains` without much hassle.

**Review by Rexicon226 — APPROVED** [review]
LGTM

**Review by Rexicon226 — APPROVED** [review]
LGTM 2.0

**Pull request merged**
